internal/controllers/machine/machine_controller_noderef.go

```diff
@@ -131,6 +136,46 @@ func (r *Reconciler) reconcileNode(ctx context.Context, cluster *clusterv1.Clust
 	return ctrl.Result{}, nil
 }
 
+func reconcileManagedLabels(machine *clusterv1.Machine, node *corev1.Node) bool {
+	// Reconcile managed Labels.
+	var labelsHaveChanged bool
+	desiredLabels := getManagedLabels(machine.Labels)
+	currentNodeLabels := getManagedLabels(node.Labels)
+
+	// Reconcile desired labels.
+	for desiredKey, desiredValue := range desiredLabels {
+		if currentValue, ok := currentNodeLabels[desiredKey]; !ok || desiredValue != currentValue {
+			node.Labels[desiredKey] = desiredValue
+			labelsHaveChanged = true
+		}
+	}
+
+	// Delete current labels which are not desired anymore.
+	for currentKey := range currentNodeLabels {
+		if _, ok := desiredLabels[currentKey]; !ok {
+			labelsHaveChanged = true
+			delete(node.Labels, currentKey)
+		}
+	}
+
+	return labelsHaveChanged
+}
+
+// getManagedLabels gets a map[string]string and returns another map[string]string
+// filtering out labels not managed by CAPI.
+func getManagedLabels(labels map[string]string) map[string]string {
+	managedLabels := make(map[string]string)
+	for key, value := range labels {
+		for _, managedPrefix := range clusterv1.ManagedNodeLabelPrefixes {
```

Comment:
If I got it right with this approach we are continuously reconciling all the managed prefixes, but according to the proposal only one of them should be reconciled and the other two should be applied once.

Reply:
Using ssa now in order to ensure that Cluster API will manage only the subset of labels coming from Machine objects and ignores labels applied to Nodes concurrently by other users/other components.